include/circt/Support/Scheduling/Scheduler.h

```diff
@@ -0,0 +1,207 @@
+
+#ifndef CIRCT_SUPPORT_SCHEDULING_SCHEDULER_H
+#define CIRCT_SUPPORT_SCHEDULING_SCHEDULER_H
+
+#include "circt/Support/LLVM.h"
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SetVector.h"
+
+namespace circt {
+namespace sched {
+
+/// This class models the most basic scheduling problem.
+///
+/// A problem instance is comprised of:
+///
+///  - *Operations*: The vertices in the data-flow graph to be scheduled.
+///    Using `mlir::Operation*` is the obvious mapping, but other data types can
+///    be used as the scheduler treats them as opaque handles. We use the
+///    abbreviation `op` to mean operation (handle) in the scheduling API.
+///  - *Dependences*: The edges in the data-flow graph to be scheduled, modeling
+///    a precedence relation between the involved operations. Individual operand
+///    and result indices can be distinguished. We use the abbreviation `dep` to
+///    in the scheduling API.
+///  - *Operator types*: An abstraction of the characteristics of the target
+///    representation, e.g. modules in the  HLS component library, available
+///    functional units, etc. -- operations are executed on instances/units of
+///    an operator type. Again, these are opaque handles for the scheduler. We
+///    use `opr` to denote operator type handles in the API.
+///
+/// All components of the problem (ops, deps, oprs, as well as the problem
+/// itself) can be annotated with *properties*. In this basic problem, we model
+///
+/// - `assocOpr`, an operation-property that stores the operator type associated
+///   with each operation.
+/// - `latency`, an operator type-property denoting the number of time steps
+///   after which the operator's result is available.
+/// - `startTime`, an operation-property for the time step in which an operation
+///   is started. Together, the start times for all operations represent the
+///   problem's solution, i.e. the schedule.
+///
+/// Subclasses, i.e. base classes for more complex scheduling problems as well
+/// as concrete algorithms, can declare additional properties as needed. The
+/// top-level entrypoint for algorithms is the pure virtual `schedule()` method.
+//
+/// The `check...` methods perform validity checks before scheduler, e.g. all
+/// operation handles referenced in dependences are registered, all ops have an
+/// associated operator type, etc.
+///
+/// The `verify...` methods check the correctness of the solution determined by
+/// a concrete scheduling algorithm, e.g. that there are start times available
+/// for each registered operation, and the precedence constraints as modeled by
+/// the dependences are satisfied.
+class Scheduler {
+  //===--------------------------------------------------------------------===//
+  // Aliases for API types
+  //===--------------------------------------------------------------------===//
+public:
+  /// Opaque handle for operations
+  using OperationHandle = void *;
```

Comment:
Could this just be an `Operation *`? That's a common way to opaquely refer to operations, and I think it works in all the APIs here.

Reply:
Yes, I'm going to change that, I realised that just having a `void*` is terrible for emitting errors.